File: apps/client-macos/Sources/TimeTrack/Tracking/ManualNudgeMonitor.swift
import Foundation
import CoreGraphics

/// Slice 2.4 — the manual-mode (`autoStartOnLogin=false`) nudge poller. One 15s timer reads the
/// SAME content-free idle scalar as `WorkspaceObserver` (`CGEventSource.secondsSinceLastEventType`
/// — no key/pointer content) and, purely locally, decides between two mutually-exclusive-by-tracker
/// -state notifications, both NOTIFY-ONLY:
///   • forgot-to-start — present + not tracking for `forgotToStartSeconds` → "start?"
///   • manual idle     — a manual clock live + idle >= threshold → "still tracking?"
/// The manual-idle nudge NEVER stops the clock (a manual entry is the user's own action —
/// CLAUDE.md/design invariant): this type does not touch `TimeTracker` or `IdleMonitor`, so it
/// produces no stop and no spurious IdleEvent. No network, no logging. Installed only behind
/// `AckGate` on the online `!ackRequired` path (see AppDelegate).
final class ManualNudgeMonitor {
    private let notifier: LocalNotifying
    private let idleThresholdSeconds: Int
    private let forgotToStartSeconds: Int
    private let isTracking: () -> Bool
    private let isPaused: () -> Bool
    private let pollInterval: TimeInterval
    /// Auto mode: present the forgot-to-start reminder as a VISIBLE window instead of posting a
    /// notification. Auto-tracking failing to start is a fault, not a gentle hint, and the
    /// notifier silently drops everything on a build macOS never authorized (see `LocalNotifier`).
    /// `nil` (the default) keeps the manual-mode behaviour: a plain notification.
    private let presentForgotToStart: ((_ title: String, _ body: String) -> Void)?
    /// Auto mode passes `false`: `IdleMonitor` already owns the idle nudge there (`idle-nudge`),
    /// and a second "still tracking?" for the same idle stretch would just compete with it.
    private let emitsManualIdleNudge: Bool
    /// True while an away keep/discard prompt is on screen awaiting the user. They are not
    /// tracking and they are present, which is exactly the forgot-to-start shape — but the
    /// prompt IS the thing asking them to act, so a second window on top of it is noise.
    private let isAwaitingResolution: () -> Bool

    private var activeSince: Date?
    private var firedForgot = false
    private var firedManualIdle = false
    private var timer: Timer?

    init(notifier: LocalNotifying, idleThresholdSeconds: Int, forgotToStartSeconds: Int,
         isTracking: @escaping () -> Bool, isPaused: @escaping () -> Bool,
         pollInterval: TimeInterval = 15,
         presentForgotToStart: ((_ title: String, _ body: String) -> Void)? = nil,
         emitsManualIdleNudge: Bool = true,
         isAwaitingResolution: @escaping () -> Bool = { false }) {
        self.notifier = notifier
        self.idleThresholdSeconds = idleThresholdSeconds
        self.forgotToStartSeconds = forgotToStartSeconds
        self.isTracking = isTracking
        self.isPaused = isPaused
        self.pollInterval = pollInterval
        self.presentForgotToStart = presentForgotToStart
        self.emitsManualIdleNudge = emitsManualIdleNudge
        self.isAwaitingResolution = isAwaitingResolution
    }

    func start() {
        timer?.invalidate()
        let t = Timer(timeInterval: pollInterval, repeats: true) { [weak self] _ in self?.sample() }
        RunLoop.main.add(t, forMode: .common)
        timer = t
    }

    func stop() {
        timer?.invalidate()
        timer = nil
        activeSince = nil
        firedForgot = false
        firedManualIdle = false
    }

    private func sample() {
        let idle = CGEventSource.secondsSinceLastEventType(.combinedSessionState,
                                                           eventType: .init(rawValue: ~0)!)
        tick(idleSeconds: Int(idle), now: Date())
    }

    /// Pure decision logic (timer-free) — the tested surface.
    func tick(idleSeconds: Int, now: Date) {
        if isTracking() {
            // Manual clock live → manual-idle nudge only. Forgot-to-start is meaningless here.
            activeSince = nil
            firedForgot = false
            if idleSeconds >= idleThresholdSeconds {
                if !firedManualIdle, emitsManualIdleNudge {
                    let minutes = max(1, Int((Double(idleSeconds) / 60.0).rounded()))
                    notifier.notify(id: "manual-idle", title: "Time tracking",
                                    body: "Idle for \(minutes) min — still tracking?")
                    firedManualIdle = true
                }
            } else {
                firedManualIdle = false   // active again → re-arm
            }
            return
        }

        if isPaused() || isAwaitingResolution() {
            // Mid-pause manual session — no nudges.
            activeSince = nil
            firedForgot = false
            firedManualIdle = false
            return
        }

        // Not tracking, not paused → forgot-to-start.
        firedManualIdle = false
        if idleSeconds >= idleThresholdSeconds {
            activeSince = nil          // user is away → break the stretch
            firedForgot = false        // re-arm for the next presence
            return
        }
        if activeSince == nil { activeSince = now }
        if let since = activeSince, !firedForgot,
           now.timeIntervalSince(since) >= Double(forgotToStartSeconds) {
            let title = "Time tracking"
            let body = "You've been active \(forgotToStartSeconds / 60) min without tracking — start?"
            if let presentForgotToStart {
                presentForgotToStart(title, body)
            } else {
                notifier.notify(id: "forgot-to-start", title: title, body: body)
            }
            firedForgot = true
        }
    }
}

File: apps/client-windows/src/NiftyTimer/Tracking/ManualNudgeMonitor.cs
using System.Globalization;
using NiftyTimer.Notifications;

namespace NiftyTimer.Tracking;

/// <summary>
/// The manual-mode (<c>autoStartOnLogin = false</c>) nudge decider. It reads the SAME content-free
/// idle scalar as <see cref="SessionObserver"/> and, purely locally, chooses between two
/// mutually-exclusive-by-tracker-state notifications, both NOTIFY-ONLY:
///
/// <list type="bullet">
///   <item><b>forgot-to-start</b> — present and not tracking for <c>forgotToStartSeconds</c> → "start?"</item>
///   <item><b>manual idle</b> — a manual clock is live and idle ≥ threshold → "still tracking?"</item>
/// </list>
///
/// The manual-idle nudge NEVER stops the clock (CLAUDE.md §1 — a manual entry is the user's own
/// action). This type holds no reference to <see cref="TimeTracker"/> or any monitor, only two
/// predicates, so it structurally cannot produce a stop or a spurious IdleEvent.
///
/// Auto mode installs a second instance as its own forgot-to-start reminder, as the macOS client
/// does: <c>presentForgotToStart</c> raises a window instead of a balloon (a stopped clock in auto
/// mode is a fault, not a hint), <c>emitsManualIdleNudge</c> is off because the auto coordinator
/// already owns the idle nudge, and <c>isAwaitingResolution</c> stands it down while an away prompt
/// is already asking the person to act.
/// </summary>
public sealed class ManualNudgeMonitor
{
    private readonly ILocalNotifier _notifier;
    private readonly int _idleThresholdSeconds;
    private readonly int _forgotToStartSeconds;
    private readonly Func<bool> _isTracking;
    private readonly Func<bool> _isPaused;
    private readonly Action<string, string>? _presentForgotToStart;
    private readonly bool _emitsManualIdleNudge;
    private readonly Func<bool> _isAwaitingResolution;

    private DateTimeOffset? _activeSince;
    private bool _firedForgot;
    private bool _firedManualIdle;

    public ManualNudgeMonitor(
        ILocalNotifier notifier,
        int idleThresholdSeconds,
        int forgotToStartSeconds,
        Func<bool> isTracking,
        Func<bool> isPaused,
        Action<string, string>? presentForgotToStart = null,
        bool emitsManualIdleNudge = true,
        Func<bool>? isAwaitingResolution = null)
    {
        _notifier = notifier;
        _idleThresholdSeconds = idleThresholdSeconds;
        _forgotToStartSeconds = forgotToStartSeconds;
        _isTracking = isTracking;
        _isPaused = isPaused;
        _presentForgotToStart = presentForgotToStart;
        _emitsManualIdleNudge = emitsManualIdleNudge;
        _isAwaitingResolution = isAwaitingResolution ?? (static () => false);
    }

    public void Reset()
    {
        _activeSince = null;
        _firedForgot = false;
        _firedManualIdle = false;
    }

    /// <summary>The pure decision logic — timer-free, and the tested surface.</summary>
    public void Tick(int idleSeconds, DateTimeOffset now)
    {
        if (_isTracking())
        {
            // A manual clock is live → the manual-idle nudge only. Forgot-to-start is meaningless.
            _activeSince = null;
            _firedForgot = false;

            if (idleSeconds < _idleThresholdSeconds)
            {
                _firedManualIdle = false; // active again → re-arm
                return;
            }

            if (!_firedManualIdle && _emitsManualIdleNudge)
            {
                var minutes = AwayMinutes.Of(idleSeconds);
                _notifier.Notify(
                    "manual-idle",
                    "Time tracking",
                    string.Create(CultureInfo.InvariantCulture, $"Idle for {minutes} min — still tracking?"));
                _firedManualIdle = true;
            }

            return;
        }

        if (_isPaused() || _isAwaitingResolution())
        {
            // Mid-pause manual session, or an away prompt already asking them to act — no nudges
            // of either kind.
            Reset();
            return;
        }

        // Not tracking, not paused → forgot-to-start.
        _firedManualIdle = false;

        if (idleSeconds >= _idleThresholdSeconds)
        {
            _activeSince = null;  // the person is away → break the stretch
            _firedForgot = false; // re-arm for the next presence
            return;
        }

        _activeSince ??= now;

        if (!_firedForgot && now - _activeSince.Value >= TimeSpan.FromSeconds(_forgotToStartSeconds))
        {
            const string title = "Time tracking";
            var body = string.Create(
                CultureInfo.InvariantCulture,
                $"You've been active {_forgotToStartSeconds / 60} min without tracking — start?");

            if (_presentForgotToStart is { } present)
            {
                present(title, body);
            }
            else
            {
                _notifier.Notify("forgot-to-start", title, body);
            }

            _firedForgot = true;
        }
    }
}

/// <summary>
/// Adapts <see cref="ManualNudgeMonitor"/> to the signal edge, so it rides the same poller as
/// everything else rather than starting a second timer over the same idle scalar.
///
/// Sleep and lock are treated as an immediate long idle rather than being ignored. A machine that
/// has just woken has plainly not been "active without tracking", so the forgot-to-start stretch
/// has to break — otherwise closing the lid for an hour would be indistinguishable from working.
/// </summary>
public sealed class NudgeSignalAdapter : ISignalReceiver
{
    private readonly ManualNudgeMonitor _monitor;
    private readonly int _awaySeconds;
    private readonly Func<DateTimeOffset> _clock;

    public NudgeSignalAdapter(
        ManualNudgeMonitor monitor,
        int awaySeconds,
        Func<DateTimeOffset>? clock = null)
    {
        _monitor = monitor;
        _awaySeconds = awaySeconds;
        _clock = clock ?? (() => DateTimeOffset.UtcNow);
    }

    public void Tick(int idleSeconds) => _monitor.Tick(idleSeconds, _clock());

    public void MarkAway() => _monitor.Tick(_awaySeconds, _clock());

    public void Resume() => _monitor.Reset();
}
